Rescale with featurewise normalization bug fix

hadipash opened the pull request:
### Summary
Featurewise centering, normalization and ZCA whitening work incorrectly with the `rescale` parameter. This PR fixes the bug.

### Related Issues
#100 

### PR Overview
- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

rragundez commented on keras_preprocessing/image/image_data_generator.py:
```diff
@@ -943,6 +943,9 @@ def fit(self, x,
             np.random.seed(seed)
 
         x = np.copy(x)
+        if self.rescale:
```
Even though somewhat obvious can you add a comment explaining why rescaling before featurewise normalization, etc.. ?

hadipash replied:
In the case of featurewise normalization and ZCA whitening rescaling isn’t really necessary since they apply their own kind of "scaling". So, this bug fix is needed more for consistent behavior no matter whether the `rescale` parameter is specified or not (before the code will produce unexpected results if the `rescale` parameter is specified).

The change to the code is useful when a user wants to apply rescaling with featurewise centering only. With this bug fix the code will produce correct outputs.

rragundez replied:
I meant a comment in the code, explaining to the developer why rescaling before...

rragundez commented on tests/image/image_data_generator_test.py:
```diff
@@ -442,5 +442,34 @@ def test_random_transforms():
+            featurewise_std_normalization=True)
+        generator.fit(images)
+        batch = generator.flow(images, batch_size=8).next()
+        assert abs(np.mean(batch)) < 1e-3
```
please make it a bit tighter 1E-6

hadipash replied:
Even without my changes to the code I couldn’t produce outputs with constant precision higher than 1e-6. Mostly it is in the range of 1e-4 to 1e-7 (tried on both Ubuntu and Windows).

rragundez replied:
that's a bit weird, I would expect that precision level. Any ideas why or from where is the difference coming from?

hadipash replied:
I found out why this happened. The reason is that even if I pass `dtype='float64'` to the `ImageDataGenerator`, it creates an iterator (`NumpyArrayIterator`, `DirectoryIterator`, or `DataFrameIterator`) using the default `dtype='float32'` thus decreasing the precision. I fixed this problem in the latest commit and set precision in tests to 1e-6.

rragundez commented on tests/image/image_data_generator_test.py:
```diff
@@ -442,5 +442,34 @@ def test_random_transforms():
+            zca_whitening=True)
+        generator.fit(images)
+        batch = generator.flow(images, batch_size=8).next()
+        assert np.max(np.abs(batch)) <= 1
```
please check thee mean here as well.

hadipash replied:
I digged deeper into the ZCA whitening topic and proposed a better test condition in the latest commit. Basically, after applying whitening, we obtain a new matrix **_Y_** (**_Y_** = **_WX_**, where **_W_** is a decorrelating matrix) where **_YY_**<sup>T</sup> is a diagonal matrix such that **_YY_**<sup>T</sup> = _(n-1)_**_I_** (_n_ is a number of data samples).

(More about this can be found in [Learning Multiple Layers of Features from Tiny Images](http://www.cs.toronto.edu/~kriz/learning-features-2009-TR.pdf) in Appendix A The ZCA whitening transformation)

hadipash commented:
Thanks @rragundez for your help! Few comments on the latest commit. I followed your suggestions on enhancing test cases but in the case of `featurewise_std_normalization` it's impossible to achieve 1e-6 precision because on the lines [722](https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/image_data_generator.py#L722) and [967](https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/image_data_generator.py#L967) of `image_data_generator.py` X is divided by the standard deviation + 1e-6 (`x /= (self.std + 1e-6)`), so the precision is smaller. Therefore, I set the level to 1e-5.

Additionally, I added `dtype` argument for iterators initialization so that data can be treated as other types rather than only `float32` (as explained above).

Finally, I introduced a better test case for ZCA whitening (also explained above).

rragundez commented on keras_preprocessing/image/image_data_generator.py:
```diff
@@ -910,6 +913,9 @@ def fit(self, x,
         Only required if `featurewise_center` or
         `featurewise_std_normalization` or `zca_whitening` are set to True.
 
+        In case when `rescale` is set to a value, rescaling is applied to
```
~In case~ When `rescale` ...
